docs/RFCS/20210604_distributed_token_bucket.md

```diff
@@ -0,0 +1,416 @@
+assigned a fraction of the global fill rate. This happens via API calls through
+the KV connector (see `kvtenant.Connector`). Each call is implemented as a
+transaction involving the system table.
+
+The global bucket keeps track of cumulative RU usage, which is reported
+periodically (on a per-tenant basis) to Prometheus.
+
+## Distributed bucket algorithm
+
+We describe the algorithm in terms of "nodes" (a tenant's SQL pods)
+communicating with a centralized ("global") token bucket. The tokens are RUs.
+Note that we describe a single instance of the algorithm; in reality there is a
+separate instance for each tenant.
+
+The nodes periodically request tokens from the global bucket. The amount of
+requested tokens is adjusted based on the current usage, with the goal of a
+target global request period. The requests are sent preemptively, a short time
+before the current tokens are estimated to run out.
+
+The global bucket operates in two modes:
+ - if there are sufficient "burst" tokens already in the bucket, the tokens are
+   granted immediately;
+ - otherwise, a fraction of the global refill rate is granted to the node and
+   the tokens are granted over a period according to this refill rate. Fewer
+   tokens can be granted so that the period does not exceed the target global
+   request period.
+
+Each node adjusts its local token bucket accordingly, either adding burst tokens
+or setting up a refill rate.
+
+The global refill rate is distributed among the nodes using a shares-based
+system. Each node calculates a share value and the global bucket maintains the
+sum of the current share values. The fraction of rate granted to a node is the
+ratio between that node's shares and the total shares. The shares reflect the
+load at each node (more details in the shares calculation section).
+
+### Global bucket state
+
+At any point in time, the global bucket maintains these values:
+ - the current amount of tokens (RUs); this value can be negative. It is
+   immediately adjusted when tokens are granted to the nodes (even when they are
+   granted over a time period).
+ - the current sum of shares
+ - the total amount of tokens used (since the tenant was created); this value
+   reflects RUs that were used with certainty. In abnormal cases (like node
+   crashes), the error is always on the side of under-counting.
+
+### Request input and output values
+
+We describe the input and output values of a request to the global bucket.
+
+Input:
+ - previous shares value for the node;
+ - new shares value for the node;
+ - tokens requested;
+ - information about RU consumption by the node since last request.
+
+Output:
+ - timestamp of when the bucket update became effective (used to account for
+   shares decay, see below);
+ - tokens granted (at most the amount requested);
+ - optional granted timestamp; if set, it indicates that the tokens cannot be
+   used right away, but they must "trickle" over a time period ending at this
```

Comment:
No, it can only use them "smoothed" out over the time interval. It is basically giving the tenant a local refill rate (which should be a fraction of the global refill rate).

Reply:
Ok, consider just saying explicitly where it gets that rate then to take the guesswork out of it.